source/h2_stream.c

```diff
@@ -1054,3 +1197,136 @@ struct aws_h2err aws_h2_stream_on_decoder_rst_stream(struct aws_h2_stream *strea
 
     return AWS_H2ERR_SUCCESS;
 }
+
+static int s_stream_write_data(
+    struct aws_http_stream *stream_base,
+    const struct aws_http2_stream_write_data_options *options) {
+    struct aws_h2_stream *stream = AWS_CONTAINER_OF(stream_base, struct aws_h2_stream, base);
+    struct aws_h2_connection *connection = s_get_h2_connection(stream);
+
+    /* queue this new write into the pending write list for the stream */
+    struct aws_h2_stream_data_write *pending_write =
+        aws_mem_calloc(stream->base.alloc, 1, sizeof(struct aws_h2_stream_data_write));
+    pending_write->data_stream = aws_input_stream_acquire(options->data);
+    pending_write->end_stream = false;
+    bool schedule_cross_thread_work = false;
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(stream);
+        {
+            if (stream->synced_data.api_state != AWS_H2_STREAM_API_STATE_ACTIVE) {
+                s_unlock_synced_data(stream);
+                s_stream_data_write_destroy(stream, pending_write, AWS_ERROR_INVALID_STATE);
+                AWS_LOGF_ERROR(
+                    AWS_LS_HTTP_STREAM,
+                    "Cannot write DATA frames to an inactive or closed stream, stream=%p",
+                    (void *)stream_base);
+                return aws_raise_error(AWS_ERROR_INVALID_STATE);
+            }
+
+            if (stream->synced_data.manual_write_ended) {
+                s_unlock_synced_data(stream);
+                s_stream_data_write_destroy(stream, pending_write, AWS_ERROR_INVALID_STATE);
+                AWS_LOGF_ERROR(
+                    AWS_LS_HTTP_STREAM,
+                    "Cannot write DATA frames to a stream after end, stream=%p",
+                    (void *)stream_base);
+                /* Fail with error, otherwise, people can wait for on_complete callback that will never be invoked. */
+                return aws_raise_error(AWS_ERROR_INVALID_STATE);
+            }
+            pending_write->on_complete = options->on_complete;
+            pending_write->user_data = options->user_data;
+
+            aws_linked_list_push_back(&stream->synced_data.pending_write_list, &pending_write->node);
+            schedule_cross_thread_work = !stream->synced_data.is_cross_thread_work_task_scheduled;
+            stream->synced_data.is_cross_thread_work_task_scheduled = true;
+        }
+        s_unlock_synced_data(stream);
+    } /* END CRITICAL SECTION */
+
+    if (schedule_cross_thread_work) {
+        AWS_H2_STREAM_LOG(TRACE, stream, "Scheduling stream cross-thread work task");
+        /* increment the refcount of stream to keep it alive until the task runs */
+        aws_atomic_fetch_add(&stream->base.refcount, 1);
+        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &stream->cross_thread_work_task);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_stream_end_stream_get_length(struct aws_input_stream *stream, int64_t *length) {
```

Comment:
Well, the input_stream_new_from cursor doesn't have the ownership of the buffer. It will just do a copy of cursor... So, maybe we still need this fake input_stream

Reply:
still think we can get away with using aws_input_stream_new_from_cursor()